src/scripts/upgrades/Upgrade.ts

```diff
@@ -0,0 +1,79 @@
+///<reference path="../GameConstants.ts"/>
+///<reference path="Cost.ts"/>
+///<reference path="CostFactory.ts"/>
+
+class Upgrade {
+    name: any;
+    displayName: string;
+    maxLevel: number;
+    _level: KnockoutObservable<number> = ko.observable();
+
+    // Describes whether this upgrade increases or decreases a number.
+    // (e.g. power is increasing, time is decreasing).
+    increasing: boolean;
+
+    // Optional array of costs
+    costList: Cost[] = [];
+    // Optional array of benefits
+    bonusList: number[] = [];
+
+    constructor(name: any, displayName: string, maxLevel: number, costList: Cost[], bonusList: number[], increasing = true) {
+        this.name = name;
+        this.displayName = displayName;
+        this.maxLevel = maxLevel;
+        this.level = 0;
+        this.costList = costList;
+        this.bonusList = bonusList;
```

Comment:
Is it necessary (or even possible) to have a way to make sure that `costList.length >= maxLevel && bonusList.length >= maxLevel + 1` during compile?

Reply:
We could write some tests for it if we ever decide to implement those.